text/006-local-js-dependencies.md

```diff
@@ -0,0 +1,347 @@
+module. This RFC proposes using the wasm itself to pass along these objects,
+like so:
+
+```rust
+// lib.rs
+
+#[wasm_bindgen(file = "local-snippet.js")]
+extern {
+    fn take_u8_slice(memory: &JsValue, ptr: u32, len: u32);
+}
+
+#[wasm_bindgen]
+pub fn call_local_snippet() {
+    let vec = vec![0,1,2,3,4];
+    let mem = wasm_bindgen::memory();
+    take_u8_slice(&mem, vec.as_ptr() as usize as u32, vec.len() as u32);
+}
+```
+
+```js
+// local-snippet.js
+
+export function take_u8_slice(memory, ptr, len) {
+    let slice = new UInt8Array(memory.arrayBuffer, ptr, len);
+    // ...
+}
+```
+
+Here the `wasm_bindgen::memory()` existing intrinsic is used to pass along the
+memory object to the imported JS snippet. To mirror this we'll add
+`wasm_bindgen::function_table()` as well to access the function table.
+
+Eventually we may want a more explicit way to import the memory/table, but for
+now this should be sufficient for expressiveness.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* The initial RFC is fairly conservative. It doesn't work with `--nodejs` out of
+  the gate nor `--no-modules`. Additionally it doesn't support JS snippets
+  importing other JS initially. Note that all of these are intended to be
+  supported in the future, it's just thought that it may take more design than
+  we need at the get-go for now.
+
+* JS snippets must be written in vanilla ES module JS syntax. Common
+  preprocessors like TypeScript can't be used. It's unclear how such
+  preprocessed JS would be imported. It's hoped that JS snippets are small
+  enough that this isn't too much of a problem. Larger JS snippets can always be
+  extracted to an NPM package and postprocessed there. Note that it's always
+  possible for authors to manually run the TypeScript compiler by hand for these
+  use cases, though.
+
+* The relatively popular `--no-modules` flag is proposed to be deprecated in
+  favor of a `--browser` flag, which itself will have a breaking change relative
+  to today. It's thought though that `--browser` is only very rarely used so is
+  safe to break, and it's also thought that we'll want to avoid breaking
+  `--no-modules` as-is today.
+
+* JS files are imported relative to the crate root rather than the file doing
+  the importing, unlike `mod` statements in Rust. It's thought that we can't get
+  file-relative imports working with stable `proc_macro` APIs today, but if the
+  implementation can manage to finesse it this RFC would propose instead making
+  file imports relative to the current file instead of crate root.
+
+* Local JS snippets are required to be written in ES module syntax. This may be
+  a somewhat opinionated stance, but it's intended to make it easier to add
+  future features to `wasm-bindgen` while continuing to work with JS.
+
+# Rationale and Alternatives
+[alternatives]: #rationale-and-alternatives
+
+Currently there aren't any competing designs solving this problem, so there
```

Comment:
The major alternative is the `js!` macro in stdweb, which allows for writing JS snippets inline in your Rust code.

Since there isn't any importing (and no `.js` files), this makes it easier to implement: it's easy for it to support other formats like "no modules" or "CommonJS modules".

And the question about "crate relative" or "file relative" doesn't exist.

And the question about "how to import a JS file inside of a JS file" doesn't exist.

And it automatically supports transitive crate dependencies.

To be clear, I'm not actually proposing for wasm-bindgen to implement `js!`.

I think this RFC makes sense for wasm-bindgen. I just want to make it clear that there *is* an alternative.

Reply:
I think I didn't word this section well enough to convey this, I've had another go at it